server/util/snapshotSpaceCache.ts

```diff
@@ -0,0 +1,80 @@
+import { Op } from 'sequelize';
+import JobRunner from './cacheJobRunner';
+import { DB } from '../database';
+import { factory, formatFilename } from '../../shared/logging';
+import { ChainInstance } from '../models/chain';
+
+const log = factory.getLogger(formatFilename(__filename));
+
+// map of snapshot spaces and associated chains
+// ALT: chainId could be an Array<ChainInstance> if we want to preserve it
+type CacheT = { [snapshotCache: string] : Set<string> };
+
+export default class SnapshotSpaceCache extends JobRunner<CacheT> {
+  private _models: DB;
+    constructor(
+      models: DB,
+      // Update the cache every hour
+      jobTimeS = 60 * 60,
+      bitch: CacheT = { snapshotCache : new Set<string>() }
+    ) {
+      super(bitch, jobTimeS);
+      this._models = models;
+    }
+
+  public async start() {
+    super.start();
+    log.info(`Started Snapshot Space Cache.`);
+  }
+
+  public async check(snapshot : string): Promise<boolean> {
+    const result = await this.access((async (c: CacheT): Promise<boolean | undefined> => {
+      // Just in case cache is empty
+      if (!c.snapshotCache.size) {
+        // Call _job to populate cache
+        await this._job(c);
+      }
+
+      // Check if the snapshot space is in the cache
+      return c.snapshotCache.has(snapshot);
+    }));
+
+    return result;
+  }
+
+  protected async _job(c: CacheT): Promise<void> {
+    // If the cache is empty
+    if (!c.snapshotCache.size) {
+      log.info(`Populating Snapshot Space Cache.`);
+    } else {
+      log.info(`Updating Snapshot Space Cache.`);
+    }
+    // NOTE: This sucks and is not really efficient but given how the DB
+    // and CacheJobRunner are set up, this seems to be the best option
+    //    - populating and updating the cache are the same because
+    //      any sort of comparison between the cache and the DB to
+    //      filter out spaces already in the cache would involve
+    //      additional in-depth querying of the DB which defeats the
+    //      purpose of the cache
+    //    - populating the cache has to happen here and not in the
+    //      check method because cacheJobRunner makes the _job method
+    //      run before all others
+    const allSnapshots = await this._models.Chain.findAll({
+      attributes: ['snapshot'],
+      where: {
+        snapshot: {
+          [Op.ne]: []
+        }
+      },
+    });
+
+    // Get all snapshot spaces out of the ChainInstance array and add it to the snapshotCache
+    allSnapshots.forEach((chainInstance: ChainInstance) => {
+      chainInstance.snapshot.forEach((snapshotSpace: string) => {
+        c.snapshotCache.add(snapshotSpace);
```

Comment:
I don't quite understand how `add` works here, given it's a mapping of -- what -- chains to snapshot spaces? We may also want to query `id` in the above database call.

Reply:
The snapshotCache isn't a mapping, its just a list of snapshot space names (lyra.eth, amp.eth, etc.) as strings stored in a Set (to prevent duplicates), so the add is just adding the space names to the Set.

Reply:
It is a mapping based on `type CacheT = { [snapshotCache: string] : Set<string> };` -- but what is mapping to what?

Reply:
Should we be 

> I know nothing about the specifics of this PR but I took a glance at the cache and have a general suggestion. We should start looking at using the available Redis instance rather than direct in-memory caches. The server will start running into memory issues (if it isn't already) if we implement only in-memory caches. A Redis cache is extremely efficient and only adds a few ms of latency which are likely saved anways in the long run because saving in-memory takes up valuable space. Also, a Redis cache would persist between server restarts.

Seconding this